src/sync/mutex.rs

```diff
@@ -24,78 +24,59 @@ pub struct MutexGuard<'a, T: ?Sized> {
 #[derive(Debug)]
 struct MutexState {
     holder: Option<TaskId>,
-    waiters: TaskSet,
-    clock: VectorClock,
 }
 
 impl<T> Mutex<T> {
     /// Creates a new mutex in an unlocked state ready for use.
     pub const fn new(value: T) -> Self {
-        let state = MutexState {
-            holder: None,
-            waiters: TaskSet::new(),
-            clock: VectorClock::new(),
-        };
-
+        let state = MutexState { holder: None };
         Self {
-            inner: std::sync::Mutex::new(value),
             state: RefCell::new(state),
+            semaphore: BatchSemaphore::const_new(1, Fairness::Unfair),
+            inner: std::sync::Mutex::new(value),
         }
     }
 }
 
 impl<T: ?Sized> Mutex<T> {
     /// Acquires a mutex, blocking the current thread until it is able to do so.
     pub fn lock(&self) -> LockResult<MutexGuard<'_, T>> {
-        let me = ExecutionState::me();
+        let me = current::me();
 
         let mut state = self.state.borrow_mut();
-        trace!(holder=?state.holder, waiters=?state.waiters, "waiting to acquire mutex {:p}", self);
+        trace!(holder=?state.holder, semaphore=?self.semaphore, "waiting to acquire mutex {:p}", self);
+        drop(state);
 
-        // If the lock is already held, then we are blocked
-        if let Some(holder) = state.holder {
-            if holder == me {
-                panic!("deadlock! task {:?} tried to acquire a Mutex it already holds", me);
+        if !self.semaphore.is_closed() {
```

Comment:
Feels a bit strange that you're checking if the semaphore is closed here, but then you call `unwrap` on the acquire call on line 70.  If we're not expecting the semaphore to be closed, maybe just assert that here?

Reply:
This is the other way though: *if* it is not closed, then we can `unwrap` the `acquire_blocking` call. If it is closed, then we will continue and should get a poisoned lock from the inner mutex.